core/templates/components/code-mirror/codemirror-mergeview.component.ts

```diff
@@ -27,14 +27,22 @@ import { downgradeComponent } from '@angular/upgrade/static';
 export class CodemirrorMergeviewComponent implements
   AfterViewInit, OnInit, OnChanges {
   @Input() options = {};
-  @Input() leftValue;
-  @Input() rightValue;
-  codeMirrorInstance: CodeMirror.MergeView.MergeViewEditor;
+  // Below properties have been assigned as empty string if undefined.
+  // See function 'ngAfterViewInit' for reference.
+  @Input() leftValue: string | undefined;
+  @Input() rightValue: string | undefined;
+  // Non-null assertion operator(!) to ignore
+  // the case when CodeMirror is undefined or null.
+  codeMirrorInstance!: CodeMirror.MergeView.MergeViewEditor;
+  // We can't directly test for window.CodeMirror to be undefined
+  // so we are assigning it to 'originalCodeMirror' here.
+  originalCodeMirror: typeof window.CodeMirror | undefined = window.CodeMirror;
```

Comment:
I am a bit skeptical of changing code to fit the tests. As I see there are three ways to get around this. Each more "evil" than the previous option 😛 . All of them involve removing this pattern of originalCodeMirror.

1. We use windowRef here.
```
  codeMirrorInstance!: CodeMirror.MergeView.MergeViewEditor;

  constructor(
    private elementRef: ElementRef,
    private ngZone: NgZone,
    private windowRef: WindowRef
  ) { }

  ngOnInit(): void {
    // Require CodeMirror.
    if (
      (this.windowRef.nativeWindow as typeof window).CodeMirror === undefined
    ) {
      throw new Error('CodeMirror not found.');
    }
  }
```

Then in tests we spyOnProperty windowRef.
```
  it('should throw error if codemirror is undefined', () => {
    spyOnProperty(TestBed.inject(WindowRef), 'nativeWindow').and.returnValue(
      {}
    );
    expect(() => {
      component.ngOnInit();
    }).toThrowError('CodeMirror not found.');
  });
```

The second is using `as unknown as  typeof window.CodeMirror` but let's not use that. There is third (and also a fourth) but I am just not comfortable explaining bad coding patterns. If you want you can try and guess what they are and why they are not good (if you have time to spare).

Reply:
done!